internal/util/observability/funccall.go

```diff
@@ -18,52 +18,65 @@ import (
 	"context"
 	"runtime"
 	"runtime/trace"
+	"sync"
 
 	"github.com/FerretDB/FerretDB/internal/util/resource"
 )
 
 // funcCall tracks function calls.
 type funcCall struct {
-	token  *resource.Token
-	region *trace.Region
+	cancelCtx context.CancelFunc
+	token     *resource.Token
+	region    *trace.Region
 }
 
-// FuncCall adds observability to a function call.
+// FuncCall adds observability to a function or method call.
+// It returns a new derived context and its cancel function.
+//
+// FuncCall should be called at the very beginning of the observable function,
+// and the returned cancel function should be called at exit.
+// The returned context should be passed down.
 //
-// It should be called at the very beginning of the function,
-// and returned function should be called at exit.
-// The returned function must not be passed or stored.
 // The only valid way to use FuncCall is:
 //
 //	func foo(ctx context.Context) {
-//	    defer FuncCall(ctx)()
-//	    // ...
+//		ctx, cancel := FuncCall(ctx)
+//		defer cancel()
+//
+//		bar(ctx)
+//		// ...
+//	}
 //
 // For the Go execution tracer, FuncCall creates a new region for the function call
 // and attaches it to the task in the context (or background task).
-func FuncCall(ctx context.Context) func() {
+func FuncCall(ctx context.Context) (context.Context, context.CancelFunc) {
+	ctx, cancelCtx := context.WithCancel(ctx)
```

Comment:
To be honest, I don't understand the idea of having `context.WithCancel` here.

Reply:
Or, better say, I don't undertrstand why we want to return `ctx`. If a caller has a gorutine and passes this inherited context to that gorutine, we will have a problem.